### PR TITLE
Added validation for verification deadline on course publication endpoint.

### PR DESCRIPTION
**Pull request by waheedahmed**
[ECOM-2381](https://openedx.atlassian.net/browse/ECOM-2381)

Validation of course verification deadlines before attempting publication to the LMS:
1. Only courses with a verified seat (i.e., corresponding to "verified" or "professional" (with verification!) course mode) are allowed to have a verification deadline.
2. For courses with a verified seat, the verification deadline is optional.
3. If a course has both an upgrade and verification deadline set, then the verification deadline must be after the upgrade deadline.

### CODE REVIEW

**Comment by waheedahmed** [comment]
@awais786 Please review.

**Comment by waheedahmed** [comment]
@rlucioni Please review.

**Comment by tasawernawaz** [comment]
LGTM 👍

File: ecommerce/courses/publishers.py [hunk 1]
```diff
@@ -95,7 +96,18 @@ def publish(self, course, access_token=None):
         verification_deadline = self.get_course_verification_deadline(course)
         modes = [self.serialize_seat_for_commerce_api(seat) for seat in course.seat_products]
 
-        has_credit = 'credit' in [mode['name'] for mode in modes]
+        is_deadline_invalid = any(
+            [parse(mode['expires']) > course.verification_deadline if mode['expires'] else False for mode in modes]
```
Comment (rlucioni) [inline ecommerce/courses/publishers.py]: You don't need the outer brackets here.

File: ecommerce/courses/publishers.py [hunk 2]
```diff
@@ -95,7 +96,18 @@ def publish(self, course, access_token=None):
...
+        )
+        if is_deadline_invalid:
+            logger.error(
+                u'Failed to publish commerce data for [%s] to LMS. The verification deadline must occur '
+                u'AFTER the upgrade deadline.', course_id
```
Comment (rlucioni) [inline ecommerce/courses/publishers.py]: Nitpick: please lowercase "after," here and below.

File: ecommerce/courses/publishers.py [hunk 3]
```diff
@@ -118,10 +130,12 @@ def publish(self, course, access_token=None):
         data = {
             'id': course_id,
             'name': name,
-            'verification_deadline': verification_deadline,
             'modes': modes,
         }
 
+        if 'verified' in mode_names or 'professional' in mode_names:
+            data['verification_deadline'] = verification_deadline
```
Comment (rlucioni) [inline ecommerce/courses/publishers.py]: Why is this necessary? `verification_deadline` defaults to None if the course doesn't have one.

Comment (rlucioni) [inline ecommerce/courses/publishers.py]: I don't think we want the key to go missing from the returned data like this.

File: ecommerce/courses/publishers.py (hunk 1 above)
Comment (rlucioni) [inline ecommerce/courses/publishers.py]: You have a bug here if the verification deadline is `None`. Any value is greater than `None`, but that doesn't necessarily mean that the configuration is invalid.

The goal of this story is to add validation of verification deadlines, not validation of upgrade deadlines. This check should only occur if `verification_deadline` is not None. Instead of accessing `course.verification_deadline` directly, use `verification_deadline` from above. Refer to the validation logic from edx-platform for an example: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/commerce/api/v1/serializers.py#L67-L88.

Also, please name this variable more clearly: `is_verification_deadline_invalid`.

File: ecommerce/courses/publishers.py (hunk 3 above)
Comment (waheedahmed) [inline ecommerce/courses/publishers.py]: OK got you'r second point we don't want the key to go missing but for first point what if `verification_deadline` is not None for audit mode? Like you mentioned on the ticket in first point "Only courses with a verified seat (i.e., corresponding to "verified" or "professional" (with verification!) course mode) are allowed to have a verification deadline."

**Comment by waheedahmed** [comment]
@rlucioni does it looks good now? And can you please tell why code coverage is failing while my test is covering all my changes.

**Comment by rlucioni** [comment]
@waheedahmed I spoke with @clintonb about this. ECOM-2381 was created almost a year ago, before verification deadlines were validated by serializers on the LMS. A lot has changed since then. We don't need multiple points of validation. That being the case, this PR and the corresponding ticket should be closed.

@schenedx please make sure that the team re-verifies the need for tickets older than three months before starting to work on them.
